### PR TITLE
SI-6943 warn on value class miscomparison.

### PR DESCRIPTION
**Pull request by paulp**
There's a very dangerous situation running around when you
combine universal equality with value classes:

  // All over your code
  val x = "abc"
  if (x == "abc") ...

  // Hey let's make x a value class
  val x = new ValueClass("abc")
  // Uh-oh

There was until now no warning when comparing a value class
with something else. Now there is.

### CODE REVIEW

**Comment by paulp** [comment]
Review @retronym

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 1]
```diff
@@ -1122,6 +1125,11 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         // unused
         def possibleNumericCount = onSyms(_ filter (x => isNumeric(x) || isMaybeValue(x)) size)
         val nullCount            = onSyms(_ filter (_ == NullClass) size)
+        def isNonsenseValidClassCompare = (
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: "Valid"? Or "Value"?

**Comment by retronym** [comment]
LGTM.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 1 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: "NonsenseValid" is how I like to say invalid. It's just a thing.

**Pull request merged**